src/sentry/api/endpoints/project_rules.py

```diff
@@ -44,66 +48,219 @@ def pre_save_rule(instance, sender, *args, **kwargs):
+        }
+        keys_to_check.update(self.SPECIAL_FIELDS)
+
+        return keys_to_check
+
+    def _get_func_to_call(self, key_to_check: str) -> Callable:
+        return self._matcher_funcs_by_key.get(key_to_check, self._default_matcher)
+
+    def _default_matcher(self, existing_rule: Rule, key_to_check: str) -> MatcherResult:
+        """
+        Default function that checks if the key exists in both rules for comparison, and compares the values.
+        """
+        match_results = MatcherResult()
+
+        existing_rule_key_data = existing_rule.data.get(key_to_check)
+        current_rule_key_data = self._rule_data.get(key_to_check)
+        if existing_rule_key_data and current_rule_key_data:
+            match_results.has_key = True
+
+        if match_results.has_key:
+            match_results.key_matches = existing_rule_key_data == current_rule_key_data
+        return match_results
+
+    def _environment_matcher(self, existing_rule: Rule, key_to_check: str) -> MatcherResult:
+        """
+        Special function that checks if the environments are the same.
+        """
+
+        # Do the default check to see if both rules have the same environment key, and if they do, use the result.
+        if (
+            base_result := self._default_matcher(existing_rule, key_to_check)
+        ) and base_result.has_key:
+            return base_result
+
+        # Otherwise, we need to do the special checking for keys
+        match_results = MatcherResult()
+        if self._rule:
+            if existing_rule.environment_id and self._rule.environment_id:
+                # If the existing rule and our rule both have environment ids, check if it's the same
+                match_results.has_key = True
+                match_results.key_matches = (
+                    existing_rule.environment_id == self._rule.environment_id
+                )
+            elif (
+                existing_rule.environment_id
+                and not self._rule.environment_id
+                or not existing_rule.environment_id
+                and self._rule.environment_id
+            ):
+                # Otherwise, if one of the rules has an environment key, but the other does not, the key was checked,
+                # but it is obviously not the same anymore
+                match_results.has_key = True
+        else:
+            current_rule_key_data = self._rule_data.get(key_to_check)
+            if existing_rule.environment_id and current_rule_key_data:
+                match_results.has_key = True
+                match_results.key_matches = existing_rule.environment_id == current_rule_key_data
+            elif (
+                existing_rule.environment_id
+                and not current_rule_key_data
+                or not existing_rule.environment_id
+                and current_rule_key_data
+            ):
+                match_results.has_key = True
```

Comment:
same

Reply:
We can't get rid of `has_key` that easily, see conversation at: https://github.com/getsentry/sentry/pull/64922/files#r1488478610